app/ix/ginas/models/converters/HTMLStringConverter.java

```diff
@@ -0,0 +1,150 @@
+package ix.ginas.models.converters;
+
+import java.io.StringReader;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.w3c.tidy.Tidy;
+
+/**
+ * Strings manipulation class
+ * for strings stored in HTML format.
+ *
+ * @author epuzanov
+ * @since 2.6
+ */
+public class HTMLStringConverter extends AbstractStringConverter {
+
+    private static String[] allowedHtmlTags = new String[] {"I", "SUB", "SUP", "SMALL"};
+    private static Pattern htmlTagPattern = Pattern.compile("<\\s*/?([^>]+)\\s*>");
+    private static char[] brackets = "()[]{}".toCharArray();
+    private static String[] htmlStrings = {"&amp;", "&larr;", "&rarr;", "&lt;", "&gt;", "&plusmn;", "-",
+        "&Alpha;", "&alpha;", "&Beta;", "&beta;", "&Gamma;", "&gamma;",
+        "&Delta;", "&delta;", "&Epsilon;", "&epsilon;", "&Zeta;", "&zeta;", "&Eta;", "&eta;",
+        "&Theta;", "&theta;", "&Iota;", "&iota;", "&Kappa;", "&kappa;", "&Lambda;", "&lambda;",
+        "&Mu;", "&mu;", "&Nu;", "&nu;", "&Xi;", "&xi;", "&Omicron;", "&omicron;", "&Pi;", "&pi;",
+        "&Rho;", "&rho;", "&Sigma;", "&sigma;", "&Tau;", "&tau;", "&Upsilon;", "&upsilon;",
+        "&Phi;", "&phi;", "&Chi;", "&chi;", "&Psi;", "&psi;", "&Omega;", "&omega;", "&szlig;",
+        "&Auml;", "&auml;", "&Ouml;", "&ouml;", "&Uuml;", "&uuml;", "&sect;", "&macr;",
+        "&#8553;&#8546;", "&#8553;&#8547;", "&#8553;&#8548;", "&#8553;&#8549;", "&#8553;&#8550;",
+        "&#8553;&#8551;", "&#8553;&#8552;", "&#8553;&#8553;", "&#8544;", "&#8545;",
+        "&#8546;", "&#8547;", "&#8548;", "&#8549;", "&#8550;", "&#8551;", "&#8552;", "&#8553;",
+        "&#8554;", "&#8555;", "<small>D</small>", "<small>L</small>",
+        "<small>N</small>", "~", "<i>", "</i>", "</sup><sup>",
+        "</sub><sub>", "<sup>", "</sup>", "<sub>", "</sub>", "&zwnj;", "&quot;"};
+    private static String[] stdStrings = {"&", "<--", "-->", "<", ">", "+/-", "-",
+        "ALPHA", "alpha", "BETA", "beta", "GAMMA", "gamma",
+        "DELTA", "delta", "EPSILON", "epsilon", "ZETA", "zeta", "ETA", "eta",
+        "THETA", "theta", "IOTA", "iota", "KAPPA", "kappa", "LAMBDA", "lambda",
+        "MU", "mu", "NU", "nu", "XI", "xi", "OMIKRON", "omikron", "PI", "pi",
+        "RHO", "rho", "SIGMA", "sigma", "TAU", "tau", "UPSILON", "upsilon",
+        "PHI", "phi", "CHI", "chi", "PSI", "psi", "OMEGA", "omega", "ss",
+        "Ae", "ae", "Oe", "oe", "Ue", "ue", "Par.", "XFF", "XIII", "XIV",
+        "XV", "XVI", "XVII", "XVIII", "XIX", "XX", "I", "II",
+        "III", "IV", "V", "VI", "VII", "VIII", "IX", "X",
+        "XI", "XII", "D", "L", "N", "-", "", "", "", "", "(", ")", "<", ">", "", "\""};
+    private static Tidy tidy;
```

Comment:
Is JTidy threadsafe?  is there harm making a new instance each time?

Reply:
JTidy is not thread-safe. We can make a new tidy instance within validationErrors method.